### PR TITLE
Add Fn.Param.is_comptime * Fix std.sort.block

### PR DESCRIPTION
**Pull request by iacore**
This patch feels very flaky. I hope it doesn't break other things.

Closes https://github.com/ziglang/zig/issues/16454.

### CODE REVIEW

File: src/Sema.zig [hunk 1]
```diff
@@ -16515,10 +16515,9 @@ fn zirTypeInfo(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Ai
                     .val = if (is_generic) .none else param_ty,
                 } });
 
-                const is_noalias = blk: {
-                    const index = std.math.cast(u5, i) orelse break :blk false;
```
Comment (iacore) [inline src/Sema.zig:16519]: `orelse break :blk false` is very sus here. I think this may be a bug.

**Comment by iacore** [comment]
bootstrap chain:

- zig 0.11.0 + [d2cf2e3](https://github.com/ziglang/zig/pull/17211/commits/d2cf2e3ea1df3245aae687bee83bb34ad7aefbd9) -> zig1
- zig1 + [11d1b0c](https://github.com/ziglang/zig/pull/17211/commits/11d1b0caabc0e2b7541c23818fa3e10383a9cfec) -> zig2

File: src/Sema.zig (hunk 1 above)
Comment (mlugg) [reply src/Sema.zig:16519]: No, this logic is correct - our compiler implementation only allows the first 32 parameters to be marked `noalias`. (But the new approach is a bit cleaner anyway.)

Comment (iacore) [reply src/Sema.zig:16519]: The new approach simply crashes. I don't think it's cleaner.

Comment (mlugg) [reply src/Sema.zig:16519]: Oh whoops, I didn't see that it was `@intCast` to `u5`. You still need to handle that case manually. Best would probably be:
```
const is_noalias = if (std.math.cast(u5, i)) |idx| func_ty_info.paramIsNoalias(idx) else false;
```
...and same for `is_comptime`.

**Comment by iacore** [comment]
Can someone disable CI? I don't think CI can test this change.

**Comment by mlugg** [comment]
CI is required to pass - it's running the standard bootstrap chain. You need to add a commit that generates zig1 at the right time to preserve functionality of the bootstrap. (Andrew will recreate this commit before merging for security reasons.)

**Comment by mlugg** [comment]
Wait, it's not even a zig1 issue - your code is literally just wrong. There's no `Params` field in the function info - it's `params`.

**Comment by iacore** [comment]
> Wait, it's not even a zig1 issue - your code is literally just wrong. There's no `Params` field in the function info - it's `params`.

oops. I think I gave the wrong commits

**Comment by iacore** [comment]
How do I use the native x86 backend? The one without LLVM.

File: lib/std/sort/block.zig [hunk 2]
```diff
@@ -103,15 +103,17 @@ pub fn block(
...
-            }
-        }.lessThan else struct {
+        // if (is_ctx_comptime) struct {
+        //     fn lessThan(comptime ctx: @TypeOf(context), lhs: T, rhs: T) bool {
+        //         const lt = lessThanFn(ctx, lhs, rhs);
+        //         const gt = lessThanFn(ctx, rhs, lhs);
+        //         std.debug.assert(!(lt and gt));
+        //         return lt;
+        //     }
+        // }.lessThan else
```
Comment (mlugg) [inline lib/std/sort/block.zig:115]: Why have you commented this logic out? Is it not the whole point of this PR?

Comment (iacore) [reply lib/std/sort/block.zig:115]: It's added back in another patch.

**Comment by iacore** [comment]
yay CI passes somewhat

[Test file I used to make this patch.](https://github.com/iacore/zig-issue-16454) I don't know if the last commit works for other occasions.

**Comment by mlugg** [comment]
> yay CI passes somewhat.

I'm confused at what you're referring to here - I don't see any passes so far?

I don't understand what you mean by "stage 1" and "stage 2". As mentioned above, this change works fine as a single change to the compiler - zig1 has not and need not be modified. To be clear, splitting your work across different commits is having zero effect on the CI runs - every run is completely distinct from others.

This entire PR can and should be squashed into a single commit. Or perhaps two, the first adding `is_comptime` - fixing any test failures that arise - and the second fixing the std function.

**Comment by iacore** [comment]
I made it into two commits.

**Comment by mlugg** [comment]
Ah, yes, that's my bad - apologies, I got confused. This change *does* need a zig1 update, because the existing zig1 will expect the standard library to not contain the `is_comptime` field when generating zig2.

You need your first commit to introduce the handling for the `is_comptime` field into Sema, but *not* to add it to std; then have a commit that updates zig1 (use `zig build update-zig1`); then have the commit that adds `is_comptime`.

Of course, the `std.sort` change is fine to have in one commit afterwards.

**Comment by iacore** [comment]
> You need your first commit to introduce the handling for the `is_comptime` field into Sema, but _not_ to add it to std; then have a commit that updates zig1 (use `zig build update-zig1`); then have the commit that adds `is_comptime`.

I can't understand this. Do I need 3 commits without the commit for `std.sort.block`?.

**Comment by mlugg** [comment]
Yes, so 4 total commits:
* Make Sema set the `is_comptime` field when using `@typeInfo`
* Update zig1
* Add `is_comptime` field into `std.builtin.Type.Fn.Param`
* Handle comptime parameters in `std.sort.block`

**Comment by iacore** [comment]
how do I make a commit that update zig1?

**Comment by mlugg** [comment]
`zig build update-zig1`

**Comment by iacore** [comment]
Is this fine now?

**Comment by mlugg** [comment]
Looks about right - ignore the `x86_64-windows` CI failing, that's just because you're missing a few commits from master.

**Comment by andrewrk** [comment]
Closing abandoned pull request. CI checks failing, last activity more than 30 days ago.

**Pull request closed without merging**